### PR TITLE
Enable Signed Header Auth for Data Collector, and Configure the Data Collector Automatically

### PR DESCRIPTION
**Pull request by danielsdeleo**
### Description
- Allow the Chef Automate data collector to authenticate using signed headers with the client key.
- Attempt to automatically configure the data collector when running in client-server mode
- Token auth still available for either mode if configured
- Check for the presence of both the data collector URL and token to enable the collector in solo mode.
### Issues Resolved

Internal Chef Ticket
### Check List
- [x] New functionality includes tests
- [x] All tests pass
- [x] RELEASE_NOTES.md, has been updated if required (not required for bugfixes, required for API changes)
- [x] All commits have been signed-off for the Developer Certificate of Origin. See https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco

### CODE REVIEW

File: chef-config/lib/chef-config/config.rb [hunk 1]
```diff
@@ -857,7 +857,7 @@ def self.windows_home_path
       # Full URL to the endpoint that will receive our data. If nil, the
       # data collector will not run.
       # Ex: http://my-data-collector.mycompany.com/ingest
-      default :server_url,       nil
+      default(:server_url) { File.join(config_parent.chef_server_url, "/data_collector") }
```
Comment (stevendanna) [inline chef-config/lib/chef-config/config.rb]: Chef server currently has a minimal proxy config for the data collector at `/data-collector`

https://github.com/chef/chef-server/blob/master/omnibus/files/private-chef-cookbooks/private-chef/templates/default/nginx/nginx_chef_api_lb.conf.erb#L48-L57

**Comment by danielsdeleo** [comment]
@sersut I added a bunch of logging to the data collector enable/disable decision-making. For case (2) I decided not to add anything because the initial data collector POST will fail with an informative error message that includes the configured URL, which should be enough information for a customer to figure out the problem.

**Comment by sersut** [comment]
👍

**Comment by danielsdeleo** [comment]
Note: I am planning to do end-to-end tests for a few scenarios once the server piece is done, so please consider this PR on hold for now.

File: spec/unit/data_collector_spec.rb [hunk 2]
```diff
@@ -23,6 +23,14 @@
 require "chef/resource_builder"
 
 describe Chef::DataCollector do
+
+  # TODO:
```
Comment (srenatus) [inline spec/unit/data_collector_spec.rb]: Are these still open? (I'm not sure what exactly you tracked here, but it seems like these have been taken care of by this change, doesn't it?)

File: chef-config/lib/chef-config/config.rb [hunk 3]
```diff
@@ -857,7 +857,7 @@ def self.windows_home_path
       # Full URL to the endpoint that will receive our data. If nil, the
       # data collector will not run.
       # Ex: http://my-data-collector.mycompany.com/ingest
-      default :server_url,       nil
+      default(:server_url) { File.join(config_parent.chef_server_url, "/data-collector") }
```
Comment (srenatus) [inline chef-config/lib/chef-config/config.rb]: Present clients will have this one configured, right? (I.e., they're not using the default value.)

Comment (danielsdeleo) [inline chef-config/lib/chef-config/config.rb]: Yes, before this patch is applied, you must explicitly configure this in order for the data collector to be enabled (it's in the old behavior of `DataCollector.register_reporter?()` ).

File: spec/unit/data_collector_spec.rb (hunk 2 above)
Comment (danielsdeleo) [inline spec/unit/data_collector_spec.rb]: Good catch, forgot to grep my diff for `TODO`. Will remove.

**Comment by danielsdeleo** [comment]
@chef/client-core This should be G2G. My team will be doing end-to-end testing for a few scenarios including not having Automate/Chef Server Authenticated proxy available.

**Pull request merged**